\documentclass[11pt]{article}

\usepackage{fullpage}
\usepackage[numbers]{natbib}
\usepackage{amssymb}
\usepackage{amsmath}
\usepackage{bm}
\usepackage{bbm}
\usepackage{graphicx}
\usepackage{color}
\usepackage{hyperref}

\DeclareMathOperator{\erf}{erf}
\DeclareMathOperator{\lngamma}{\ln\!\Gamma}
\DeclareMathOperator{\Tr}{Tr}

% NOTE
% one sentence per line for nice git diffs
% WD: I'm a fan of initialed source comments like this




% TODO
% - ancient african substructure buzz, bantu VS san, admixture signal?
% - gbGC and isochores: https://www.annualreviews.org/doi/abs/10.1146/annurev-genom-082908-150001?casa_token=EImkIpqbayAAAAAA:Rxc8y0KLVRbZ6UfBTy3VxBRPdt3IZf7CAdVk57DfBUpvpSvuP4uHpDe9XHHCMHFBbYDxcZyFR8t0tA
% - thinking on identifiability chat with Stilianos at UO: although we allow the different mutation rates (for the different triplet contexts) to vary in time, it might be reasonable to enforce that the total mutation rate (added over the the triplets) is still rather constant in time. Without this, it’s true that we can not identify 1.\ stretching of the coalescent tree in some time interval due to a population crash from 2.\ a period of increased total mutation rate. We could enforce this with a strong spline penalty, or strictly, using a prox to a simplex.
% - inverse problem intro, forward Vs inverse problems, and ill-posedness. Point out \eta is nonlinear, but \mu is linear
% - tidy up the math with some lemmas and theorems


\title{Population genetic inference of mutation spectrum evolution}

% WD: the affil numbers are a little out of order, but it's nice to concisely bundle all the UW depts
\author{
W.S. DeWitt$^{1,4}$, K. Decker Harris$^{2,3}$, and K. Harris$^{1,4}$\\
{\small
$^1$Department of Genome Sciences,
$^2$Paul G.\ Allen School of Computer Science \& Engineering,}\\
{\small and $^3$Department of Biology, University of Washington, Seattle, WA}\\
{\small $^4$Computational Biology Program, Fred Hutchinson Cancer Research Center, Seattle, WA}
% \small{$^\ast$ Equal contribution}
}

\begin{document}

\maketitle

\begin{abstract}

Models in evolutionary genetics typically assume that mutation rate is constant over time and between populations and closely related species.
However, recent work casts doubt on this assumption in human and ape populations, and reveals that mutation is a complex and dynamic process.
Whether arising from variation in replication fidelity, life history, or environmental exposures, mutation rate evolution can be accompanied by changes to the \emph{mutation spectrum}: the mutation rate in different local nucleotide contexts.
We extend theoretical tools based on Kingman's coalescent to accommodate a richly parameterized mutation process, varying in time and in spectrum.
% WD: revise below when we settle on what results will be in the paper
% WD: may also want to emphasize the hypothesis generating aspect of having the mutation spectrum history
We infer human mutation spectrum histories from patterns of modern genomic diversity, allowing us to reconstruct trajectories of mutation spectrum divergence between populations, track a transient mutation spectrum perturbation through multiple populations, and characterize how mutation spectrum histories are structured by local nucleotide context.
Mutation rate and effective population size together determine the strength of genetic drift shaping genomic variation.
We introduce fast nonparametric inference of mutation spectrum history and demographic history from unphased genomes.

\end{abstract}


\section*{Introduction}\label{sec:intro}

% WD: should this be less humany?
Population genetics has proved to be a powerful lens on human history \cite{Harris2017-cd}, revealing a complex past of migration \cite{?}, replacement \cite{?}, and admixture among human groups and extinct homonids \cite{?}, and dramatic population size changes during and after range expansions \cite{?}.
Germline mutation events in the history of a population give rise to variant alleles, and demographic histories are recorded in patterns of segregation within and between current-day populations.
Genetic drift is shared when populations mix, and it is private during periods of isolation.
The strength of drift is universally influenced by the population size history, thus population genetic methods for demographic inference---expressed in the framework of Kimura's diffusion or Kingman's coalescent---attempt to recover the effective population size history $N(t)$ (where time is measured in years or generations before the present).

% WD: These examples seem a bit arbitrary, and maybe it's odd to put the more recent Lynch work next to classics like Kimura and Haldane in this paragraph
Mutation drives the evolution of populations and species, and mutation rate evolution has a storied history in the theoretical population genetics literature.
Haldane developed equilibrium theory for alleles in mutation-selection balance, and used this to provided the first principled estimate of the human mutation rate by studying hemophilia incidence \cite{Haldane2004-ov, Nachman2004-gh}.
Kimura later considered how genetic modifiers of the mutation rate---\emph{mutator alleles}---tune the optimal mutation rate to balance adaptive response to environmental changes, increasing genetic load, and the tendency of mutators to escape their deleterious consequences via recombination \cite{Kimura1967-ku}.
The \emph{drift-barrier hypothesis} of Lynch et al.\ considers the effect of genetic drift on mutation rate optima, finding that effective population size determines a limit for evolutionary optimization toward high replication fidelity, as the efficiency of selection against mutator alleles increases with $N$ \cite{Lynch2016-ff}.

The process of germline mutation is the writing mechanism that records signatures of demographic events in genomes, so its influence on modern genome variation is similar in importance to the demographic histories themselves.
Demographic inference methods can model complex population splits, migration, and admixture, in addition to modeling $N(t)$ as a piecewise constant (or exponential) function of time into the past.
Mutation, however, receives a relatively impoverished treatment: usually a single mutation rate parameter $\mu$ is understood to apply at all loci in all individuals at all times, and may be regarded as a nuisance parameter needed for time calibration of models whose natural dimensionless time is measured in Wright-Fisher generations.

Growing evidence indicates that germline mutation is a dynamic process on both interspecific and population time scales, and a complex function of polymorphic replication machinery, life history, mutagenic exposures, and genomic content.
Mutation rates among great apes have declined along the lineage leading to humans---a phenomenon termed the \emph{homonoid slowdown} \cite{Goodman1985-xc, Scally2012-rb}---, showing that mutation rate evolution between species distorts phylogenetic time calibration.
At the level of single generations, children of older parents receive more germline mutations, especially so from older fathers, likely arising from replicative errors in spermatogenesis that add $\sim 1$ additional expected mutation per year of paternal age.
% WD: parental age or paternal age? Both are found in the lit
This \emph{parental age effect} \cite{} shows that sex-specific life history traits can influence the population mutation process.
% WD: need an example of mutagenic/spontaneous germline mutation
% WD: for some of these, it is unknown if driven by replicative Vs spontaneous mutation \cite{Moorjani2016-fx, Gao2016-qp}, and that baboon paper gets at this (Molly)

A complex and polymorphic mutation process also reveals itself in associations with genomic position and local nucleotide context.
The rate of \texttt{C$\rightarrow$T} transitions is elevated at methylated CpGs due to spontaneous deamination \cite{Hwang2004-bq, Segurel2014-ss}.
GC-biased gene conversion (gbGC) refers to the tendency of stronger-binding GC alleles to overwrite AT alleles during homologous recombination \cite{Galtier2007-eb,Duret2009-wn}.
This biased non-Mendelian segregation pattern is tantamount to selection for \emph{weak to strong} mutations.
% WD: we might want to talk about isochores here, but I've yet to find a coherent definition

It is difficult to disentangle past changes in mutation rate from past changes in effective population size.
However, the myriad mechanisms that drive the mutation process suggest that its evolution may be revealed by studying the \emph{mutation spectrum}: the mutation rate in different local nucleotide contexts.
Indeed, Harris et al.\ \cite{Harris2015-wi, Harris2017-fw} used a triplet context parameterization borrowed from the somatic mutation signature literature \cite{Alexandrov2013-oe, Helleday2014-xb}, and counted single nucleotide variants in each triplet mutation type as an estimate of the time-averaged mutational input from each individual's history.
Human triplet spectra distinctly clustered according to continental ancestry group, and evidence of historical pulses in mutation activity (or suppression of repair) was found in the distribution of allele frequencies in certain mutation types.
Mathiesen et al.\ studied similar mutational signatures in triplet spectra of rare human variants \cite{Mathieson2017-rc}, and clarified alternative non-mutational hypotheses for their origin, including population differences in demography, patterns of selection, recombination, or recombination-associated processes such as gene conversion.
Rare variants in large human cohorts serve as a proxy for recent de novo mutations, and reveal mutation signatures of replication timing, recombination, and sex differences in repair processes \cite{Carlson2018-rc, Agarwal2019-dn}.

Emerging from the literature is a picture of a mutation process evolving within and between populations, anchored to genomic features, and accented by spectra of local nucleotide context.
If probabilistic models of population genetic processes are to keep pace with phenomenology, mutation deserves much more rich description in these models.
% WD: remember, "nonparametric" is just a screwy way of saying "the most parameterized"
In this paper, we build on classical theoretical tools to introduce fast nonparametric inference of population-level \emph{mutation spectrum history} (MUSH)---the relative mutation rate history in different local nucleotide contexts---alongside inference of demographic history from unphased genomes.
% WD: The thing I came for: the wreck and not the story of the wreck. The thing itself and not the myth.
In doing so, we promote mutation spectrum evolution from phenomenological statistics on modern variation to an historical object to be inferred, on the same footing as demography.

The sample frequency spectrum (SFS)---the distribution of derived allele frequencies among sampled haplotypes---is a well-studied population genetic summary statistic that is sensitive to demographic history.
Here, we extend a coalescent framework for demographic inference from the SFS to accommodate inference of the MUSH from an SFS that is resolved into different local $k$-mer nucleotide context---a more rich summary statistic that we call the $k$-SFS, where e.g.\ $k=3$ for triplet context.
We show that inference of the MUSH is a linear inverse problem.
While challenging to invert due to ill-posedness, this stands in stark contrast to the deep problems of non-identifiability that afflict demographic inference from the SFS \cite{Myers2008-jp, Bhaskar2014-fu, Terhorst2015-xt, Baharian2018-np, Rosen2018-bb} (although an estimate of the latter is also required).
We develop regularization methods, grounded in recent advances in proximal gradient optimization, to seek non-negative solutions with smoothness (under various norms) and sparsity in mutation signatures.
We infer both demographic history and MUSH nonparametrically using regularization, rather than selecting a model with a small number of constant or exponential epochs.
Our open-source software, termed \texttt{mushi} (mutation spectrum history inference), is available at \url{https://github.com/harrispopgen/mushi}.

An inferred MUSH is a richly-parameterized object to which various biological questions about evolution of the mutation process may be addressed, both intrapopulation (patterns within a single MUSH) and interpopulation (comparisons between MUSHs).
% WD: hold off on summary of results until we settle on what to show
We use \texttt{mushi} to\dots

% WD: need to decide where to put this
% -------
\section*{Non-identifiability of demographic history and total mutation rate}

As shown in appendix \ref{sec:appendix:forward:pcsws} eqn~\eqref{eqn:d2}, effective population size $\eta(t)$ and the total mutation rate $\mu(t)$ are non-identifiable at all times.
This is demonstrated by introducing a change of variables that measures time in expected number of coalescent events since the present.
This non-identifiability is similarly manifest by a change of variables to measure time in expected mutations \citetext{Peter Ralph, priv.\ comm.}.
This can be understood intuitively: an excess of variants of a given frequency can be explained by an historical bottleneck (which expands coalescent lines in the bottleneck time interval), but equally well by a period of increased total mutation rate with no demographic change.
% -------

\section*{Model}\label{sec:model}

The setting for our modeling is Kingman's coalescent \cite{Kingman1982-ge, Kingman1982-tf, Kingman1982-ys, Kingman2000-jr}, with all the usual niceties: neutrality, infinite sites, linkage equilibrium, and panmixia.
In Appendix \ref{sec:appendix} we detail our retracing of the derivation by \cite{Griffiths1998-qf} of the frequency distribution of a derived allele conditioned on the demographic history, while generalizing to a time inhomogeneous mutation process.
We make use of the results of \cite{Polanski2003-kg} and \cite{Polanski2003-ll} to facilitate computation, and build on the notation of \cite{Rosen2018-bb} for a finite-dimensional approximation of our demographic and mutation intensity histories.
Complete derivation of formulae are found in the Appendix.

Given $n$ sampled haplotypes and a demographic history $\eta(t)$ (with time measured retrospectively from the present), we show in Appendix \ref{sec:appendix:xi} that the expected SFS $\boldsymbol \xi = (\xi_1, \xi_2,\dots, \xi_{n-1})$ is a linear transform of the mutation intensity history $\mu(t)$:
\begin{equation}
  \label{eqn:transform}
\boldsymbol \xi = \mathcal{L}_{n,\eta(t)}\mu(t),
\end{equation}
where $\mathcal{L}_{n,\eta(t)}$ is a finite-rank linear operator that maps infinite-dimensional mutation intensity histories to $(n-1)$-dimensional SFS vectors.
It is dependent on $n$ and $\eta$, which we take to be fixed.

% Need to talk about this fig \ref{fig:model}.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{figures/model}
\caption{Schematic of the marked Poisson process model with $n=4$.
We condition on coalescent times $T_4=t_4,T_3=t_3,T_2=t_2$ and consider mutation intensity function $\mu(t)$.
Red dots indicate mutation events placed by time $t$, genomic position $x$, and coalescent line (which are depicted as extruded in the genomic coordinate axis, grey sheets).
The probability that a differential element $dxdt$ on a given sheet contains a mutation is proportional to the instantaneous mutation intensity $\mu(t)$.
}
\label{fig:model}
\end{figure}

In previous work \cite{Kelley, cancer, etc}, mutation spectra have been parameterized by grouping variants into mutation types according to local $k$-mer nucleotide context, and the number of variants called in each mutation type is taken as a proxy integrating an individual's ancestral mutational input in each type.
With $k$ odd and the variant registered in the central position there are $\kappa_k=2\times3\times4^{k-1}$ mutation types after collapsing by strand symmetry.
For example, there are $96$ triplet mutation types ($k=3$).
Here, we use the same $k$-mer mutation type parameterization to promote the $(n-1)$-element expected SFS vector $\xi$ to the $(n-1)\times\kappa_k$ expected $k$-SFS matrix $\Xi$.
Similarly, the mutation intensity history function $\mu(t)$ is promoted to the $\kappa_k$-element mutation spectrum history $\boldsymbol\mu(t)$, with each element giving the mutation intensity history function for one mutation type.
Equation \eqref{eqn:transform} becomes
\begin{equation}
  \label{eqn:transform_matrix}
\Xi = \mathcal{L}_{n,\eta(t)}\boldsymbol\mu^\intercal(t),
\end{equation}

For numerical implementation we consider finite-dimensional approximations of $\eta(t)$ and $\boldsymbol\mu(t)$ as piecewise constant functions of time on $m$ common epochs $[t_0, t_1), [t_1, t_2),\dots, [t_{m-1}, t_m)$ where $0=t_0 < t_1 < \dots < t_{m-1} < t_m=\infty$.
We take the epoch boundaries as fixed parameters, and in practice make them dense so as to approximate infinite-dimensional histories.
Let $\boldsymbol y = (y_1,\dots,y_m)$ denote the constant population size $\eta(t)$ during each epoch, and let the $m\times\kappa_k$ matrix $Z$ denote the constant mutation intensity during each epoch (rows) for each mutation type (columns).
In Appendix \ref{sec:appendix:pcsws} we show the linear transform \eqref{eqn:transform_matrix} reduces to the matrix equation
\begin{equation}
\label{eqn:transform_discrete}
\Xi = L_{n, \boldsymbol y} Z,
\end{equation}
where the $(n-1)\times m$ matrix $L_{n, \boldsymbol y}$ is fixed given a fixed demographic history $\boldsymbol y$.

In linkage equilibrium the log likelihood function of the mutation spectrum history $Z$ given an observed $k$-SFS matrix $X$ and demographic history $\boldsymbol y$ is given by the Poisson random field approximation \cite{?}
\[
\ell(Z; X, \boldsymbol y, n) = \Tr(X^\intercal\log(L_{n, \boldsymbol y} Z)) - \|L_{n, \boldsymbol y} Z\|_1,
\]
where $\log(\cdot)$ operates elementwise, and we've dropped a constant term wrt $Z$.

% WD: more inverse problem intro
The inverse problem \eqref{eqn:transform_discrete} is ill-posed in general,
so many very different histories can be equally consistent with the data
\cite{oscillation paper? Yun's other papers?}.
We deal with this problem using regularization, to enforce well-behaved histories.
Our cost function to minimize, a penalized log-likelihood, is
\begin{equation}
\label{eqn:penalized}
C(Z)
%\tilde\ell(\boldsymbol z)
= -\ell(Z) + R(Z) .
\end{equation}
The function $R(Z)$ incorporates our regularization.
In order to recover smooth histories,
we penalize the $L^p$ norms
of the time derivative of $\boldsymbol\mu(t)$
\[
\sum_{i=1}^{\kappa_k}\left\| \frac{d \mu_i(t)}{d t} \right\|_p^p
= \sum_{i=1}^{\kappa_k}\int_0^\infty\left|\frac{d\mu_i(t)}{dt}\right|^p dt,
\]
for $p=1,2$.
With piecewise constant histories as in \eqref{eqn:transform_discrete} this becomes
\[
\left\|\Delta Z \right\|_p^p.
\]
where $\Delta$ denotes the first difference matrix.
When $p = 1$, this is referred to as a fused LASSO or total variation (TV) penalty,
whereas $p=2$ is called a spline penalty.
We have implemented three different types of regularization.

The first option we call ``soft'' rank regularization, with
\begin{equation}
  \label{eq:regularization_soft}
  R_\mathrm{soft}(Z)
  =
  \frac{  \lambda_\mathrm{spline} }{2} \left\|\Delta Z\right\|_2^2
  +
  \lambda_\mathrm{rank} \| Z \|_*
  +
   \frac{\lambda_\mathrm{ridge}}{2}  \| Z \|_\mathrm{F}^2
   .
\end{equation}
The spline term controls the smoothness of solutions,
and the rank term allows us to recover low-rank solutions.
The rank penalty is the nuclear norm $\| Z \|_*$,
which is equivalent to the $\ell^1$-norm on the singular values.
If $Z = U S V^\intercal$ is the SVD of $Z$, with $S = \mathrm{diag}(\sigma)$ then $\| Z \|_* = \| \sigma \|_1$.
The parameters $\lambda_\mathrm{spline}$ and $\lambda_\mathrm{rank}$ control the overall strength
of smoothing and rank regularization.
The function $R_\mathrm{soft}$ is convex; therefore we have a convex optimization problem
in this scenario.

The second option is ``hard'' rank regularization, where
\begin{equation}
  \label{eq:regularization_soft}
  R_\mathrm{hard}(Z)
  =
  \frac{\lambda_\mathrm{spline}}{2} \left\|\Delta Z\right\|_2^2
  +
  \lambda_\mathrm{rank} \, \mathrm{rank}(Z)
  +
  \frac{\lambda_\mathrm{ridge}}{2} \| Z \|_\mathrm{F}^2
  .
\end{equation}
Here, the rank penalty is the actual rank of $Z$,
and note that $\mathrm{rank}(Z) = \| \sigma \|_0$.
In this case the problem is {\em not} convex.
However, we may still compute a proximal map for the non-differentiable rank term
and optimize with prox-gradients.

The third option allows more general smoothing but ignores all rank penalization:
\begin{equation}
  \label{eq:regularization_soft}
  R_\mathrm{indep}(Z)
  =
  \lambda_\mathrm{TV}
  \left\|\Delta Z\right\|_1
  +
  \frac{\lambda_\mathrm{spline}}{2} \left\|\Delta Z\right\|_2^2
  +
  \frac{\lambda_\mathrm{ridge}}{2} \| Z \|_\mathrm{F}^2
  .
\end{equation}
The TV terms control the smoothness of solutions,
and the rank term allows us to recover low-rank solutions.
The function $R_\mathrm{indep}$ is convex.
Note also that the columns of $Z$ are uncoupled,
so that we may solve this case as $\kappa_k$ different subproblems.


\section*{Results}\label{sec:results}

% Todo:
% \begin{itemize}
% \item repeat 1KG analysis like in \cite{Harris2017-fw}, see if we recapitulate Kelley's simulation-based pulse results
% \item cluster triplet time series to see if we pull in minor components.
% \end{itemize}

\begin{figure}
  \centering
  \includegraphics[width=.7\textwidth]{figures/fit_teaser}
  \caption{Simulating and inverting a pulse.}
  \label{}
\end{figure}

\subsection*{Tempora incognita: observability toward the coalescent horizon}\label{sec:model:loss}

Todo: SVD on $L_{n, \boldsymbol y}$

Intuitively, we know the SFS can't contain any information about the history beyond the TMRCA $T_2$, since mutations that occurred before then will not be segregating in the sample.
Thus, we will find it useful to penalize complexity in the history more heavily at times that are more probably ancestral to the TMRCA.

From \eqref{eqn:pi} and \eqref{eqn:r} in the Appendix \ref{sec:appendix}, the CDF of $T_2$ is
\begin{align}
F_2(t) &= 1 - \sum_{j=2}^n A_{2,j}r_2(t)\\
&= 1 - \sum_{j=2}^n A_{2,j}r_2(t)
\end{align}

\section*{Discussion}\label{sec:discussion}

% WD: Peter's references: https://www.bibsonomy.org/user/peter.ralph/mutation_spectrum

1KG cell line artifact \cite{Anderson-Trocme2019-fy} (they also use strict mask)

Although methods exist for inferring locus-wise mutation rates \cite{Bhaskar2015-bd, Nelson2012-fr}, these require very large study sizes, and do not accommodate evolution of the rate or spectrum of the mutation process per se.

New methods that approximate the ancestral recombination graphs...

We have to be very clear about how what we’re offering is different than what can be done with Relate \cite{Speidel2019-ox}.
\begin{enumerate}
\item We can deal with unphased genomes.
\item Our method allows the total mutation rate to be variable whereas Relate fixes it to a constant, and does posthoc counting.
\item I do not quite understand how they count mutations in a given epoch, since they are not mapped to specific times, only to branches.
Unless mu(t) is constant, uniform sampling on the branch is not consistent
\item How does pulse timing compare?
\end{enumerate}


\section*{Methods}\label{sec:methods}

\subsection*{Implementation and pipeline}\label{sec:methods:tool}

The \texttt{mushi} software package is available at \url{https://github.com/harrispopgen/mushi}.
% WD: cite whatever packages we use, e.g. jax, proxtv, msprime, stdpopsim, etc.

\section*{Acknowledgements}\label{sec:ack}

Peter Ralph and Andy Kern and the Kern/Ralph colab, Harris lab, Stilianos Louca, Erick Matsen, Aleksandr Aravkin, Jeff Spence

\bibliographystyle{plainnat}
\bibliography{refs}


\appendix
\input{appendix}


\end{document}
