{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "Quickstart\n",
    "==\n",
    "\n",
    "This page demonstrates how to use the Python module `mushi` to infer demographic history and mutation spectrum history"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "We use `mushi` to infer effective population size history and history of the mutation process, which we can think of as the mutation rate function over time for each of many mutation types.\n",
    "In `mushi`, we use coalescent theory and optimization techniques to learn about this history from the $k$-SFS: a matrix whose columns are sample frequency spectra (SFS) for each mutation type (or just the standard SFS if we are only interested in inferring demographic history).\n",
    "\n",
    "\n",
    "## Imports\n",
    "We first import the `mushi` package, and a few other standard packages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [],
   "source": [
    "import mushi\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading sample frequency spectrum data\n",
    "\n",
    "The most important class in the `mushi` package is the `kSFS`, which holds site frequency spectrum (SFS) data, or $k$-SFS data for multiple mutation types.\n",
    "There are several ways to instantiate a `kSFS` object (for details read the API documentation):\n",
    "\n",
    "### 1. From a `numpy` array\n",
    "\n",
    "Here we demonstrate a `kSFS` instantiated from a numpy array, and then a plot of the resulting SFS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = 1e6 / np.arange(1, 200)\n",
    "\n",
    "ksfs = mushi.kSFS(X=x)\n",
    "\n",
    "ksfs.plot_total()\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. From a tsv file\n",
    "\n",
    "For example, suppose a file `ksfs.tsv` contains $k$-SFS data (in the format generated by the seperate `mutyper` package), or simply an SFS with sample frequencies in the first column and variant counts in the second. The `kSFS` is instantiated as\n",
    "\n",
    "```python\n",
    "ksfs = mushi.kSFS(file='ksfs.tsv')\n",
    "```\n",
    "\n",
    "### 3. By simulation\n",
    "\n",
    "We can also instantiate by only specifying the number of number of sampled haplotypes $n$, then generating simulated SFS or $k$-SFS data by specifying a demographic history and a optionally a mutation spectrum history.\n",
    "We will proceed with this approach for this demonstration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ksfs = mushi.kSFS(n=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now define a true demographic history for simulation (note that for inference on real data, this is not necessary).\n",
    "In `mushi`, the demographic history is denoted by the haploid effective population size history $\\eta(t)$ (\"eta\" of t), and can be constructed using the class `mushi.eta`.\n",
    "\n",
    "**Note that for diploids, $\\eta(t)=2N(t)$.**\n",
    "\n",
    "We instantiate demographic history with a vector of times (into the past) when the population size changes, and a vector of population sizes in each interval.\n",
    "Here we define a simple bottleneck history, and plot it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "t = np.logspace(0, 4, 100)\n",
    "y = 1e4 * np.ones(t.size + 1)\n",
    "y[40:60] /= 4\n",
    "\n",
    "eta_true = mushi.eta(t, y)\n",
    "\n",
    "eta_true.plot()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we simulate SFS data using this demographic history.\n",
    "We also specify a constant mutation rate (mutations per genome per generation), a rate of ancestral misidentification, and a random seed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mu0 = 50\n",
    "ksfs.simulate(eta_true, mu0, r=0.02, seed=1)\n",
    "\n",
    "ksfs.plot_total()\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Demographic inference\n",
    "\n",
    "With SFS data defined, we can attempt to infer demographic history, and ask if it resembles the true history we used to simulate the data.\n",
    "\n",
    "Demographic inference requires an estimate of the total mutation rate, and optional regularization penalties.\n",
    "\n",
    "### Regularization parameters\n",
    "\n",
    "- The most important regularization parameter is the trend penalty, a tuple `(k, λ)`. This indicates a trend penalty of order `k` and strength `λ`. Inference will favor piece-wise polynomial solutions, with pieces of order `k`. For example using a 0th order penalty (as we do below) will favor piece-wise constant solutions. Using a 1st order penalty would allow linear pieces, and a 2nd order penalty would allow quadratic pieces, etc. Multiple trend penalty tuples can be supplied, which will result in solutions that are a mix of pieces of different orders.\n",
    "- An additional `ridge_penalty` parameter improves convergence without noticably affecting the solution, so we recommend using a small one, typically `1e-4`.\n",
    "\n",
    "We use a single 0th order trend penalty, to favor piecewise constant solutions. The choice of which trend orders to use is a model selection choice that should be judiciously chosen, possible based on domain knowledge about the population under study.\n",
    "\n",
    "We also can specify the number of iterations (if we want more than the default), and a verbosity level to print info about progress."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ksfs.infer_eta(mu0,\n",
    "               (0, 1e2),  # <-- trend penalty\n",
    "               ridge_penalty = 1e-4,\n",
    "               max_iter=300,\n",
    "               verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now plot the fit to the SFS, as well as the inferred history alongside the true history. The inferred history is contained in the attribute `eta` of the `kSFS` object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.subplot(121)\n",
    "ksfs.plot_total()\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "\n",
    "plt.subplot(122)\n",
    "eta_true.plot(label='true')\n",
    "ksfs.eta.plot(label='inferred')\n",
    "plt.legend();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The ancestral misidentification rate was also inferred (to fit the \"smile\"), and we can check that it is near the true value used to simulate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ksfs.r"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Folded SFS\n",
    "\n",
    "For non-model organisms, it is often difficult to polarize the ancestral state of SNPs.\n",
    "The SFS can be supplied with reference polarized data, and inference can be performed using the flag `folded=True`, to use the folded SFS (aggregating each sample frequence with its complement).\n",
    "We can repeate the above inference using folded SFS inference to get similar results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ksfs.infer_eta(mu0,\n",
    "               (0, 1e2),\n",
    "               ridge_penalty = 1e-4,\n",
    "               folded=True,\n",
    "               max_iter=300, verbose=True)\n",
    "\n",
    "plt.subplot(121)\n",
    "ksfs.plot_total(folded=True)\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "\n",
    "plt.subplot(122)\n",
    "eta_true.plot(label='true')\n",
    "ksfs.eta.plot(label='inferred')\n",
    "plt.legend();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Mutation spectrum history (mush) inference\n",
    "\n",
    "We now consider data that is resolved into different mutation types (e.g. 96 triplet mutation types), and we have an SFS for each type as a separate column of a $k$-SFS matrix.\n",
    "As described above, a `kSFS` can be instantiated from a numpy array, or loaded from a tsv file.\n",
    "Here we'll again simulate one, so we first need to define a true mutation spectrum history.\n",
    "\n",
    "In `mushi`, the mush is denoted by a time dependent mutation rate history vector $\\boldsymbol\\mu(t)$ (\"mu\" of t), and can be constructed using the class `mushi.mu`.\n",
    "Each component of the vector is the mutation rate for one mutation type (e.g. one triplet type, like `AAA>ACA`).\n",
    "\n",
    "We instantiate mush with a vector of times (into the past) when the mush changes, and a _matrix_ of mutation rate vectors in each interval.\n",
    "Here we define a a 96 dimensional history where all are constant in time except two: a pulse and a monotonic ramp.\n",
    "Here again, note that this step would not be necessary for inference on real data, but is needed for this demonstration.\n",
    "\n",
    "To simplify things, we'll aggregate all the constant mutation types, so we have mutation types: `TCC>TTC`, `GAA>GGA`, and `None` (which will contain the remaining 94 mutation types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Z = np.ones((t.size + 1, 3))\n",
    "\n",
    "# pulse\n",
    "Z[40:70, 0] *= 2\n",
    "Z[:40, 0] *= 1.2\n",
    "Z[:, 0] *= 2\n",
    "\n",
    "# ramp\n",
    "Z[25:75, 1] *= np.linspace(2, 1, 50)\n",
    "Z[:25, 1] *= 2\n",
    "\n",
    "# flat\n",
    "Z[:, 2] *= 94\n",
    "\n",
    "# We model a TCC>TTC, it's revertant type GAA>GGA (for ancestral misidentification),\n",
    "# and aggregate category for the remainder of mutations\n",
    "mutation_types = ['TCC>TTC', 'GAA>GGA', None]\n",
    "\n",
    "mu_true = mushi.mu(t, Z, mutation_types)\n",
    "\n",
    "mu_true.plot(('TCC>TTC', 'GAA>GGA'), alpha=0.75, lw=3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now simulate new $k$-SFS data assuming this mush, and plot just the first two (non-constant) components.\n",
    "\n",
    "Using the flag `clr=True` when plotting the $k$-SFS shows relative mutation counts (using a compositional transform) at each time, so it makes it easier to visualize."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ksfs = mushi.kSFS(n=200)\n",
    "ksfs.simulate(eta_true, mu_true, r=0.02, seed=1)\n",
    "\n",
    "plt.subplot(121)\n",
    "ksfs.plot_total()\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "\n",
    "plt.subplot(122)\n",
    "ksfs.plot(('TCC>TTC', 'GAA>GGA'), clr=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mu0 = Z[0, :].sum()\n",
    "\n",
    "ksfs.infer_eta(mu0,\n",
    "               (0, 1e2),\n",
    "               ridge_penalty = 1e-4,\n",
    "               max_iter=300, verbose=True)\n",
    "\n",
    "plt.subplot(121)\n",
    "ksfs.plot_total()\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')\n",
    "\n",
    "plt.subplot(122)\n",
    "eta_true.plot(label='true')\n",
    "ksfs.eta.plot(label='inferred')\n",
    "plt.legend();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ksfs.infer_mush((0, 1e2), (1, 1e0), # <-- mixed trend penalties\n",
    "                ridge_penalty = 1e-4,\n",
    "                max_iter=300, verbose=True)\n",
    "\n",
    "plt.subplot(121)\n",
    "ksfs.plot(('TCC>TTC', 'GAA>GGA'), clr=True)\n",
    "\n",
    "plt.subplot(122)\n",
    "mu_true.plot(('TCC>TTC', 'GAA>GGA'), ls='--')\n",
    "ksfs.mu.plot(('TCC>TTC', 'GAA>GGA'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
