

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Quickstart &mdash; mushi  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
        <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../_static/style.css" type="text/css" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Citing mushi" href="../cite.html" />
    <link rel="prev" title="Installation" href="../install.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html">
          

          
            
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">User Guide</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Imports">Imports</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Load-k-SFS-data">Load <span class="math notranslate nohighlight">\(k\)</span>-SFS data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#3-SFS-for-1000-Genomes-FIN-population"><span class="math notranslate nohighlight">\(3\)</span>-SFS for 1000 Genomes FIN population</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Target-sizes">Target sizes</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Normalized-variant-spectra">Normalized variant spectra</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Plot-the-SFS">Plot the SFS</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Plot-the-k-SFS">Plot the <span class="math notranslate nohighlight">\(k\)</span>-SFS</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Basic-model-parameters">Basic model parameters</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Total-mutation-rate">Total mutation rate</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Generation-time">Generation time</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Discrete-time-grid">Discrete time grid</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Jointly-infer-demography-and-mutation-spectrum-history">Jointly infer demography and mutation spectrum history</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Visualizing-inferred-histories">Visualizing inferred histories</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cite.html">Citing <code class="docutils literal notranslate"><span class="pre">mushi</span></code></a></li>
</ul>
<p class="caption"><span class="caption-text">Notebooks</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="simulation_simple.html">Simple simulation</a></li>
<li class="toctree-l1"><a class="reference internal" href="observability.html">Observability toward the coalescent horizon</a></li>
</ul>
<p class="caption"><span class="caption-text">API Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../mushi.html">Public API: <code class="docutils literal notranslate"><span class="pre">mushi</span></code> package</a></li>
</ul>
<p class="caption"><span class="caption-text">Developer Documentation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../developer.html">Open source code repository</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#update-documentation">Update documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#test-notebook">Test notebook</a></li>
<li class="toctree-l1"><a class="reference internal" href="../developer.html#todo-list">Todo list</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">mushi</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
      <li>Quickstart</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/notebooks/quickstart.ipynb.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Quickstart">
<h1>Quickstart<a class="headerlink" href="#Quickstart" title="Permalink to this headline">¶</a></h1>
<p>This page demonstrates how to use the Python module <code class="docutils literal notranslate"><span class="pre">mushi</span></code> to infer mutation spectrum history and demography</p>
<p>We use <code class="docutils literal notranslate"><span class="pre">mushi</span></code> to infer history of the mutation process, which we can think of as the mutation rate function over time for each triplet mutation type. In <code class="docutils literal notranslate"><span class="pre">mushi</span></code>, we use coalescent theory and optimization techniques to learn about this history from the <span class="math notranslate nohighlight">\(k\)</span>-SFS: a matrix whose columns are sample frequency spectra (SFS) for each mutation type.</p>
<div class="section" id="Imports">
<h2>Imports<a class="headerlink" href="#Imports" title="Permalink to this headline">¶</a></h2>
<p>We first import the <code class="docutils literal notranslate"><span class="pre">ksfs</span></code> module from the <code class="docutils literal notranslate"><span class="pre">mushi</span></code> package, and a few other standard packages.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">mushi</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Load-k-SFS-data">
<h2>Load <span class="math notranslate nohighlight">\(k\)</span>-SFS data<a class="headerlink" href="#Load-k-SFS-data" title="Permalink to this headline">¶</a></h2>
<div class="section" id="3-SFS-for-1000-Genomes-FIN-population">
<h3><span class="math notranslate nohighlight">\(3\)</span>-SFS for 1000 Genomes FIN population<a class="headerlink" href="#3-SFS-for-1000-Genomes-FIN-population" title="Permalink to this headline">¶</a></h3>
<p>We load <span class="math notranslate nohighlight">\(3\)</span>-SFS data for the 1000 Genomes Finnish population, which we previously computed with the <code class="docutils literal notranslate"><span class="pre">mutyper</span></code> <a class="reference external" href="https://github.com/harrispopgen/mutyper">package</a> and saved in tabular format.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span> <span class="o">=</span> <span class="n">mushi</span><span class="o">.</span><span class="n">kSFS</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s1">&#39;../example_data/3-SFS.EUR.FIN.tsv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the population variant spectrum (summing the <span class="math notranslate nohighlight">\(k\)</span>-SFS over sample frequency)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">as_df</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">17</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">family</span><span class="o">=</span><span class="s1">&#39;monospace&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;number of variants&#39;</span><span class="p">);</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Target-sizes">
<h3>Target sizes<a class="headerlink" href="#Target-sizes" title="Permalink to this headline">¶</a></h3>
<p>We will also need the masked genome size for each mutation type, which we’ve also previously computed with <code class="docutils literal notranslate"><span class="pre">mutyper</span> <span class="pre">targets</span></code>. This defines mutational target sizes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">masked_genome_size</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;../example_data/masked_size.tsv&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="n">masked_genome_size</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;mutation type&#39;</span>

<span class="n">masked_genome_size</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">family</span><span class="o">=</span><span class="s1">&#39;monospace&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;mutational target size (sites)&#39;</span><span class="p">);</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Normalized-variant-spectra">
<h3>Normalized variant spectra<a class="headerlink" href="#Normalized-variant-spectra" title="Permalink to this headline">¶</a></h3>
<p>With this we can compute the number of SNPs per target in each mutation type. Notice the enrichment of C&gt;T transitions at CpG sites.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">normalized_hit_rates</span> <span class="o">=</span> <span class="n">ksfs</span><span class="o">.</span><span class="n">as_df</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;variant count&#39;</span><span class="p">)</span>
<span class="n">normalized_hit_rates</span><span class="p">[</span><span class="s1">&#39;target size&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">masked_genome_size</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">context</span><span class="p">])</span>
                                       <span class="k">for</span> <span class="n">context</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">normalized_hit_rates</span><span class="p">[</span><span class="s1">&#39;variant count&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;&gt;&#39;</span><span class="p">)]</span>

<span class="p">(</span><span class="n">normalized_hit_rates</span><span class="p">[</span><span class="s1">&#39;variant count&#39;</span><span class="p">]</span> <span class="o">/</span>
 <span class="n">normalized_hit_rates</span><span class="p">[</span><span class="s1">&#39;target size&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">plot</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">17</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xticks</span><span class="p">(</span><span class="n">family</span><span class="o">=</span><span class="s1">&#39;monospace&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;variants per target&#39;</span><span class="p">);</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Plot-the-SFS">
<h2>Plot the SFS<a class="headerlink" href="#Plot-the-SFS" title="Permalink to this headline">¶</a></h2>
<p>The SFS is given by summing the rows of the <span class="math notranslate nohighlight">\(k\)</span>-SFS over mutation types</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">plot_total</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Plot on log scale instead of linear</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">plot_total</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Plot-the-k-SFS">
<h2>Plot the <span class="math notranslate nohighlight">\(k\)</span>-SFS<a class="headerlink" href="#Plot-the-k-SFS" title="Permalink to this headline">¶</a></h2>
<p>Plot <span class="math notranslate nohighlight">\(k\)</span>-SFS composition as a scatter (a color for each mutation type)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">plot</span><span class="p">()</span>
</pre></div>
</div>
</div>
<p>Plot again using the centered log ratio transform option (<code class="docutils literal notranslate"><span class="pre">clr=True</span></code>) to represent compositions over mutation types in each frequency class</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">clr</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Use a heatmap to show the <span class="math notranslate nohighlight">\(k\)</span>-SFS as a matrix, colored using a compositional centralization transform. This method wraps the <code class="docutils literal notranslate"><span class="pre">clustermap</span></code> function in <a class="reference external" href="https://seaborn.pydata.org/generated/seaborn.clustermap.html#seaborn.clustermap">Seaborn</a>, and you can use keyword arguments you would use in that function. A few of note: - <code class="docutils literal notranslate"><span class="pre">figsize</span></code>: a tuple giving the width and height of the figure - <code class="docutils literal notranslate"><span class="pre">col_cluster</span></code>: if <code class="docutils literal notranslate"><span class="pre">False</span></code>, don’t cluster the columns (giving an ordinary heatmap) -
<code class="docutils literal notranslate"><span class="pre">xticklabels</span></code>: if <code class="docutils literal notranslate"><span class="pre">True</span></code>, force printing of all the mutation type labels - <code class="docutils literal notranslate"><span class="pre">robust</span></code>: if <code class="docutils literal notranslate"><span class="pre">True</span></code>, automatically set the min and max values for the color scaling - <code class="docutils literal notranslate"><span class="pre">cmap</span></code>: color map name</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">17</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="n">xticklabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">robust</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Basic-model-parameters">
<h2>Basic model parameters<a class="headerlink" href="#Basic-model-parameters" title="Permalink to this headline">¶</a></h2>
<div class="section" id="Total-mutation-rate">
<h3>Total mutation rate<a class="headerlink" href="#Total-mutation-rate" title="Permalink to this headline">¶</a></h3>
<p>To compute the total mutation rate in units of mutations per masked genome per generation, we multiply an estimate of the site-wise rate by the target size</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">μ0</span> <span class="o">=</span> <span class="mf">1.25e-8</span> <span class="o">*</span> <span class="n">masked_genome_size</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="n">μ0</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Generation-time">
<h3>Generation time<a class="headerlink" href="#Generation-time" title="Permalink to this headline">¶</a></h3>
<p>To render time in years rather than generations, we use an estimate of the generation time</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">t_gen</span> <span class="o">=</span> <span class="mi">29</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Discrete-time-grid">
<h3>Discrete time grid<a class="headerlink" href="#Discrete-time-grid" title="Permalink to this headline">¶</a></h3>
<p>We define a grid of times to represent histories on, measured retrospectively from the present in units of Wright-Fisher generations.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">t</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">logspace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">log10</span><span class="p">(</span><span class="mi">200000</span><span class="p">),</span> <span class="mi">200</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="Jointly-infer-demography-and-mutation-spectrum-history">
<h2>Jointly infer demography and mutation spectrum history<a class="headerlink" href="#Jointly-infer-demography-and-mutation-spectrum-history" title="Permalink to this headline">¶</a></h2>
<p>To infer a time-calibrated mutation spectrum history (MuSH), we need to jointly estimate the demographic history, since this defines the diffusion timescale of the coalescent process.</p>
<p>We now run the optimization, setting a few parameters to control how complicated we let the histories look. We use the <code class="docutils literal notranslate"><span class="pre">verbose=True</span></code> argument to print convergence messages.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">infer_history</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">μ0</span><span class="p">,</span> <span class="n">alpha_tv</span><span class="o">=</span><span class="mf">1e2</span><span class="p">,</span> <span class="n">alpha_spline</span><span class="o">=</span><span class="mf">3e3</span><span class="p">,</span> <span class="n">alpha_ridge</span><span class="o">=</span><span class="mf">1e-4</span><span class="p">,</span>
                   <span class="n">beta_rank</span><span class="o">=</span><span class="mf">1e1</span><span class="p">,</span> <span class="n">beta_tv</span><span class="o">=</span><span class="mf">7e1</span><span class="p">,</span> <span class="n">beta_spline</span><span class="o">=</span><span class="mf">1e1</span><span class="p">,</span> <span class="n">beta_ridge</span><span class="o">=</span><span class="mf">1e-4</span><span class="p">,</span>
                   <span class="n">tol</span><span class="o">=</span><span class="mf">1e-11</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Hopefully you agree that was fast 🏎</p>
</div>
<div class="section" id="Visualizing-inferred-histories">
<h2>Visualizing inferred histories<a class="headerlink" href="#Visualizing-inferred-histories" title="Permalink to this headline">¶</a></h2>
<p>After inferring histories, we can access them as instance attributes of the <code class="docutils literal notranslate"><span class="pre">kSFS</span></code> object: - demography: <code class="docutils literal notranslate"><span class="pre">ksfs.eta</span></code> (or <code class="docutils literal notranslate"><span class="pre">ksfs.η</span></code>) - MuSH: <code class="docutils literal notranslate"><span class="pre">ksfs.mu</span></code> (or <code class="docutils literal notranslate"><span class="pre">ksfs.μ</span></code>)</p>
<p>We can inspect these, and see they’re both object with base class <code class="docutils literal notranslate"><span class="pre">mushi.histories.history</span></code></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="nb">print</span><span class="p">(</span><span class="n">ksfs</span><span class="o">.</span><span class="n">eta</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ksfs</span><span class="o">.</span><span class="n">eta</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__bases__</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="n">ksfs</span><span class="o">.</span><span class="n">mu</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ksfs</span><span class="o">.</span><span class="n">mu</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="vm">__bases__</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</div>
<p>We’ll now check that the demography has a few features we expect in the Finnish population: the out-of-Africa bottleneck shared by all Eurasians, a later bottleneck associated with northward migration, and exponential population growth toward the present. - The plot on the left shows fit to the SFS - The plot on the right shows the inferred haploid effective population size history.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">121</span><span class="p">)</span>
<span class="n">ksfs</span><span class="o">.</span><span class="n">plot_total</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">122</span><span class="p">)</span>
<span class="n">ksfs</span><span class="o">.</span><span class="n">eta</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t_gen</span><span class="o">=</span><span class="n">t_gen</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="mf">1e3</span><span class="p">,</span> <span class="mf">1e6</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<p>Now let’s take a look at the inferred mutation spectrum history (MuSH). - The plot on the left shows the measured <span class="math notranslate nohighlight">\(k\)</span>-SFS composition (points) and the fit from <code class="docutils literal notranslate"><span class="pre">mushi</span></code> (lines) - The plot on the right shows the inferred MuSH</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">16</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">121</span><span class="p">)</span>
<span class="n">ksfs</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">clr</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">subplot</span><span class="p">(</span><span class="mi">122</span><span class="p">)</span>
<span class="n">ksfs</span><span class="o">.</span><span class="n">μ</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">t_gen</span><span class="o">=</span><span class="n">t_gen</span><span class="p">,</span> <span class="n">clr</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.75</span><span class="p">)</span>
<span class="n">ksfs</span><span class="o">.</span><span class="n">μ</span><span class="o">.</span><span class="n">plot</span><span class="p">((</span><span class="s1">&#39;TCC&gt;TTC&#39;</span><span class="p">,),</span> <span class="n">t_gen</span><span class="o">=</span><span class="n">t_gen</span><span class="p">,</span> <span class="n">clr</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlim</span><span class="p">([</span><span class="mf">1e3</span><span class="p">,</span> <span class="mf">1e6</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<p>We can also plot the MuSH as a heatmap with the y axis representing time (with an interface similar to the <code class="docutils literal notranslate"><span class="pre">kSFS.clustermap</span></code> method above)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">ksfs</span><span class="o">.</span><span class="n">μ</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">t_gen</span><span class="o">=</span><span class="n">t_gen</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">17</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="n">xticklabels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">robust</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now that you have a MuSH, you can start answering questions about mutation spectrum history!🤸‍</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../cite.html" class="btn btn-neutral float-right" title="Citing mushi" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../install.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, William DeWitt, Kameron Decker Harris, and Kelley Harris

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>